Mathlib/Algebra/PresentedMonoid/Basic.lean

```diff
@@ -0,0 +1,147 @@
+@[to_additive "`of` is the canonical map from `α` to a presented additive monoid with generators
+`x : α`. The term `x` is mapped to the equivalence class of the image of `x` in `FreeAddMonoid α`"]
+def of (rels : FreeMonoid α → FreeMonoid α → Prop) (x : α) : PresentedMonoid rels :=
+  Quotient.mk (conGen rels).toSetoid (FreeMonoid.of x)
+
+section inductionOn
+
+variable {α₁ α₂ α₃ : Type* } {rels₁ : FreeMonoid α₁ → FreeMonoid α₁ → Prop}
+  {rels₂ : FreeMonoid α₂ → FreeMonoid α₂ → Prop} {rels₃ : FreeMonoid α₃ → FreeMonoid α₃ → Prop}
+
+local notation "P₁" => PresentedMonoid rels₁
+local notation "P₂" => PresentedMonoid rels₂
+local notation "P₃" => PresentedMonoid rels₃
+
+@[to_additive (attr := elab_as_elim), induction_eliminator]
+protected theorem inductionOn {δ : P₁ → Prop} (q : P₁) (h : ∀ a, δ (mk rels₁ a)) : δ q :=
+  Quotient.ind h q
+
+@[to_additive (attr := elab_as_elim)]
+protected theorem inductionOn₂ {δ : P₁ → P₂ → Prop} (q₁ : P₁) (q₂ : P₂)
+    (h : ∀ a b, δ (mk rels₁ a) (mk rels₂ b)) : δ q₁ q₂ :=
+  Quotient.inductionOn₂ q₁ q₂ h
+
+@[to_additive (attr := elab_as_elim)]
+protected theorem inductionOn₃ {δ : P₁ → P₂ → P₃ → Prop} (q₁ : P₁)
+    (q₂ : P₂) (q₃ : P₃) (h : ∀ a b c, δ (mk rels₁ a) (mk rels₂ b) (mk rels₃ c)) :
+    δ q₁ q₂ q₃ :=
+  Quotient.inductionOn₃ q₁ q₂ q₃ h
+
+end inductionOn
+
+variable {α : Type*} {rels : FreeMonoid α → FreeMonoid α → Prop}
+
+@[to_additive]
+instance : Monoid (PresentedMonoid rels) := Con.monoid (conGen rels)
+
+@[to_additive]
+theorem mul_mk (a b : FreeMonoid α) : mk rels a * (mk rels b) = mk rels (a*b) := rfl
+
+@[to_additive]
+theorem one_def : (1 : PresentedMonoid rels) = mk rels 1 := rfl
+
+/-- The generators of a presented monoid generate the presented monoid. That is, the submonoid
+closure of the set of generators equals `⊤`. -/
+@[to_additive (attr := simp) "The generators of a presented additive monoid generate the presented
+additive monoid. That is, the submonoid closure of the set of generators equals `⊤`"]
+theorem closure_range_of (rels : FreeMonoid α → FreeMonoid α → Prop) :
+    Submonoid.closure (Set.range (PresentedMonoid.of rels)) = ⊤ := by
+  rw [Submonoid.eq_top_iff']
+  intro x
+  induction' x with a
+  induction a
+  · exact Submonoid.one_mem _
+  · rename_i x
+    exact subset_closure (Exists.intro x rfl)
+  rename_i x y hx hy
+  exact Submonoid.mul_mem _ hx hy
+
+section ToMonoid
+variable {α M : Type*} [Monoid M] (f : α → M)
+variable {rels : FreeMonoid α → FreeMonoid α → Prop}
+variable (h : ∀ a b : FreeMonoid α, rels a b →  FreeMonoid.lift f a = FreeMonoid.lift f b)
+
+/-- The extension of a map `f : α → M` that satisfies the given relations to a monoid homomorphism
+from `PresentedMonoid rels → M`. -/
+@[to_additive "The extension of a map `f : α → M` that satisfies the given relations to an
+additive-monoid homomorphism from `PresentedAddMonoid rels → M`"]
+def toMonoid : MonoidHom (PresentedMonoid rels) M :=
```

Comment:
```suggestion
def toMonoid : PresentedMonoid rels →* M :=
```
I would call this PresentedMonoid.lift instead.

Reply:
I think `lift` makes more sense here (or even `desc`).

Reply:
done!